Allow more data to be counted as empty

markstory opened the pull request:
date, and time fields should all allow `[]` as an empty value. This makes the behavior of the new methods compatible with `Validator::notEmpty()`'s definition of empty. This change also makes
`allowEmptyDate(..., false)` work as a complement to its other modes.

I ran across this issue while trying to build the `notEmptyDate()`, and `notEmptyDateTime()` methods. I was unable to get them to behave opposite of the `allowEmpty*()` methods and reject `[]`.

I'm targeting this for 3.7 as I think it is a mistake in the logic that creates inconsistent behavior with `allowEmpty()` and `notEmpty()`.

markstory committed:
Allow more data to be counted as empty

date, time, and file fields should all allow `[]` as an empty value.
This makes the behavior of the new methods compatible with
`Validator::notEmpty()`'s definition of empty. This change also makes
`allowEmptyDate(..., false)` work as a complement to its other modes.

ADmad commented on src/Validation/Validator.php:
```diff
@@ -2339,7 +2339,8 @@ protected function isEmpty($data, $flags)
         }
 
-        if ($data === [] && ($flags & self::EMPTY_ARRAY)) {
+        $arrayTypes = (self::EMPTY_ARRAY | self::EMPTY_DATE | self::EMPTY_TIME | self::EMPTY_FILE);
```
The braces are not necessary. 4.x ruleset will complain about it when merging :).

markstory committed:
Remove redundant parens

markstory committed:
Don't allow [] in file fields.

Like '', an empty array is not an acceptable empty state for a file
upload field. The `[]` value should be passed onto other validator
rules as it is likely not valid.
